contracts/ProtocolReserve/ProtocolShareReserve.sol

```diff
@@ -92,8 +247,106 @@ contract ProtocolShareReserve is Ownable2StepUpgradeable, ExponentialNoError, Re
      */
     function updateAssetsState(
         address comptroller,
-        address asset
-    ) public override(IProtocolShareReserve, ReserveHelpers) {
-        super.updateAssetsState(comptroller, asset);
+        address asset,
+        IncomeType incomeType
+    ) public override(IProtocolShareReserve) {
+        require(ComptrollerInterface(comptroller).isComptroller(), "ProtocolShareReserve: Comptroller address invalid");
+        require(asset != address(0), "ProtocolShareReserve: Asset address invalid");
+        require(
+            comptroller == corePoolComptroller ||
+                PoolRegistryInterface(poolRegistry).getVTokenForAsset(comptroller, asset) != address(0),
+            "ProtocolShareReserve: The pool doesn't support the asset"
+        );
+
+        Schema schema = Schema.TWO;
+        address vToken = IPrime(prime).vTokenForAsset(asset);
+
+        if (vToken != address(0) && comptroller == corePoolComptroller && incomeType == IncomeType.SPREAD) {
+            schema = Schema.ONE;
+        }
```

Comment:
I would encapsulate this code in one internal function like this:
```
function getSchema(address comptroller, address asset, IncomeType incomeType) internal return (Schema schema) {
    schema = Schema.TWO;
    address vToken = IPrime(prime).vTokenForAsset(asset);

    if (vToken != address(0) && comptroller == corePoolComptroller && incomeType == IncomeType.SPREAD) {
        schema = Schema.ONE;
    }
}
```

This way we have located the business rule to decide the schema to apply, in an isolated function